### PR TITLE
Include generated code in determining effective regions.

### PR DESCRIPTION
**Pull request by nimakarimipour**
This PR is built on #116

This PR enhances logic for selecting the potentially impacted regions to include generated code by processors (e.g. [Lombok](https://projectlombok.org)).

`Trackers` are used to detect potentially impacted regions for a fix. These trackers only include the regions where the element was used directly. But since some processors, will copy the annotations on the element over some generated methods, these trackers will miss the real impacted regions because of the copied annotation. This PR tries to address this issue.

Changes in this PR:
- Adds `GeneratedCodeDetector` interface in `annotator-scanner` module which is responsible for determining if a `TreePath` is pointing to a generated code.
- Adds `LombokGeneratedCodeDetector`  which implements `GeneratedCodeDetector` and can detect if a path is pointing to a `Lombok` generated code. For this feature to work properly (`lombok.addLombokGeneratedAnnotation = true` must be set)
- Adds a fifth column in the element use serialization in regions which denotes if the serialized region exists in source code or is generated by a processor. If the region exists in the source code it will have the value `"SOURCE"`, otherwise, it will have the name of the processor that created it. 
- Adds `GeneratedRegionTracker` in `annotator-core` module which is responsible for extending the list of potentially impacted regions to include users of generated code that uses the target element.
- Adds `LombokTracker` which implements `GeneratedRegionTracker` and can extend regions with other regions that Lombok has created.

To enable handling regions created by Lombok, the flag below must be passed:
```
-ardl, --activate-region-detection-lombok
```

### CODE REVIEW

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/Config.java
Comment (msridhar) [inline annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/Config.java:85]: Again use enum for return

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java
Comment (msridhar) [inline annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java:43]: use enum

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (msridhar) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:26]: I think the idea here is that for any Lombok-generated method in `regions`, we also add all _uses_ of that method to the impacted regions.  Is this the right idea?  If so, let's add some comments / Javadoc documenting the logic here.  Also we should state briefly why we think this is sufficient to get all the relevant regions, given the documented assumptions on how Lombok propagates annotations (see previous comment)

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (msridhar) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:12]: Let's add some Javadoc on this class that documents what our assumptions are in terms of how exactly Lombok propagates annotations, so we know what we are trying to handle.

**Review by msridhar — COMMENTED** [review]
I think I get the high-level idea.  I have a few comments, mostly around use of `enum`s instead of `String`s

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/Config.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/Config.java]: Thank you for the suggestion, I used enum 4bbd2e4c7539c60756522d7238db930d34bcd1d2

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayV0Adapter.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/adapters/NullAwayV0Adapter.java]: Yes exactly.

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/Region.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/Region.java]: 4bbd2e4c7539c60756522d7238db930d34bcd1d2

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/Config.java
Comment (nimakarimipour) [reply annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/Config.java]: 4bbd2e4c7539c60756522d7238db930d34bcd1d2

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java
Comment (nimakarimipour) [reply annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java]: 4bbd2e4c7539c60756522d7238db930d34bcd1d2

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:62]: f0880e2067988c53d64832202e269c34e7366588

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:46]: f0880e2067988c53d64832202e269c34e7366588

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (lazaroclapp) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:75]: This works on methods only and one hop, right? I.e. if our normal exploration adds `@Nullable` to field `Foo.bar` and `Foo` is `@Data`, our initial affected regions (_before_ this fix) will include `getBar()`. Then this chance makes sure we add all methods/fields/etc that would be affected by `getBar()` itself automatically becoming `@Nullable` due to Lombok. Correct?

This might not be enough to track all the changes to fields and methods of the inner class `Foo.Builder` or their impacted regions if the class is also annotated `@Builder`. Right?

Happy if the answer is no and we can do that other case as a _follow up_ PR, btw 🙂

**Review by lazaroclapp — COMMENTED** [review]
Did a pass. Also, again, I think the description could use a bit of an edit/proofread. It's useful, but there are a few sentences that are harder to parse than they need to be.

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/Config.java
Comment (nimakarimipour) [reply annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/Config.java]: 10559d1ee17a69f81b170f6f3179084ea9d9fd65

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java
Comment (nimakarimipour) [reply annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java]: 5a0eee109f07d37d5c2b993cb84f47a5ebad38c6

Yes, that it true, thank you for noticing this.

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java
Comment (nimakarimipour) [reply annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java:67]: I was only focusing on generated methods which can be called outside of the class and extend the potentially impacted regions. But, you are right, I don't see any harm in having a comprehensive check on the generated code of any type. Please see 3621febe383f74896a79cb75ad777a417bd25504. Used the same logic for detecting region member which already covers all cases.

File: annotator-core/src/test/resources/templates/lombok/build.gradle
Comment (nimakarimipour) [reply annotator-core/src/test/resources/templates/lombok/build.gradle:69]: Thank you for noting this. I will add these flags, in the followup PR for adding more unit tests.

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/Config.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/Config.java:316]: 94f67a76c8368165cb2286e5fe3849961d2f4b7a activated it by default.

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/CompoundTracker.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/CompoundTracker.java]: d0f30c6ca12993e073fd811245d01bc260b12985

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:75]: I looked at generated code by `@Builder` and I did not see any generated code getting reused again. But I think as discussed in the follow up PR for more tests, we can also resolve this issue if it comes up 🙂

File: annotator-core/src/test/java/edu/ucr/cs/riple/core/LombokTest.java
Comment (nimakarimipour) [reply annotator-core/src/test/java/edu/ucr/cs/riple/core/LombokTest.java:60]: As discussed offline, we decided to have these tests in a follow up PR.

**Comment by nimakarimipour** [comment]
@lazaroclapp Thank you for the review. I resolved all your comments and this is ready for another round. Thank you 🙂.

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java
Comment (msridhar) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java:30]: Brief Javadoc would be good

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (msridhar) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:42]: If I understand correctly, I think this should be clearer:
```suggestion
 * all callers of any method region in lombok generated code. This will guarantee
```

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (msridhar) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:72]: Not related to this PR, but it seems strange that the `getRegions()` method returns an `Optional<Set<Region>>` rather than just a `Set<Region>` that could be empty.  Maybe fix in a follow up

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/SymbolUtil.java
Comment (msridhar) [inline annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/SymbolUtil.java:173]: Why is there a comment about lombok here?  Feels somewhat unrelated to what this method is doing

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java
Comment (msridhar) [inline annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java:43]: I don't think we need this constant, since it's only used once.  Just return `SourceType.LOMBOK` from the method

**Review by msridhar — APPROVED** [review]
Only minor comments this time.  Approving assuming they will be addressed before landing

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/LombokTracker.java:72]: @msridhar I wanted to distinguish between trackers "not having any knowledge about the passed element" and "having data but not having any actual usage". But this was required in the old designs and should be updated in follow up PRs.

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java
Comment (nimakarimipour) [reply annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java:35]: d97d3c4660d066e421bfa5c0580d53ddd80bee9d

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/SymbolUtil.java
Comment (nimakarimipour) [reply annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/SymbolUtil.java]: Sorry thanks for noticing this. This was copied from existing method. 37387665f141cec005997acb344854d6dea6685a

File: annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java
Comment (nimakarimipour) [reply annotator-scanner/src/main/java/edu/ucr/cs/riple/scanner/generatedcode/LombokGeneratedCodeDetector.java]: 71cc44387e60b52e1f588bee27b4295934846ef7

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java
Comment (lazaroclapp) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java:42]:
```suggestion
   * @return Set of potentially impacted regions including all known potentially impacted
```

But also this is both hard to parse and - I think - inaccurate.

You are not extending the list exclusively with generated regions, but, rather whenever a region in the initial set (a region affected by a fix) is a generated region, we are also adding every region that would be impacted by a fix within the generated region, since the annotation processor might propagate changes from the original fix to this generated region independent of the annotator's own logic. Is this correct? Is there a clearer way to explain this? (cc: @msridhar )

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java
Comment (msridhar) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java:33]:
```suggestion
 * Interface for Trackers which can extend the set of potentially impacted regions to account for regions generated by an annotation processor that do not exist in source code. (e.g. <a
 * href="https://projectlombok.org">Lombok</a>)
```

File: annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java
Comment (msridhar) [inline annotator-core/src/main/java/edu/ucr/cs/riple/core/metadata/trackers/generatedcode/GeneratedRegionTracker.java:44]: I think this might work better (and we don't need to redundantly do `@param` and `@return`):
```suggestion
   * Given an input set of impacted regions, returns an extended set of potentially-impacted regions, accounting for the properties of impacted regions in code generated by an annotation processor.
   */
```
Needs to be formatted

**Review by msridhar — APPROVED** [review]

**Pull request merged**
